### PR TITLE
Add a Talk API for OCP

### LINKED ISSUE #29922: Talk OCP API
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**

As an app developer I would like to create conversations and add participants to them in the back-end of my app.

**Describe the solution you'd like**

Similar to other integrations there would be a broker service that delegates between apps and the talk app. The talk app will register itself into this.

Methods required at the time of writing
* `isAvailable(): bool`
* `createConversation(string $name, IUser[] $moderators, IUser[] $participants): IConversation`
  * IConverstation has to expose the room token and the URL

**Describe alternatives you've considered**

N/a

**Additional context**

Apps that need this
* https://github.com/nextcloud/calendar/issues/3480

---

cc @nextcloud/talk

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/server/issues/29922

### Example usage

```php

use OCP\Talk\IBroker;

class AppointmentService {

    private $talkBroker;

    public function __construct(IBroker $talkBroker) {
        $this->talkBroker = $talkBroker;
    }

    public function book(IUser $organizer, string $guest): void {
        // ...

        if ($this->broker->hasBackend()) {
            $conversation = $this->broker->createConversation(
                "Appointment with $guest",
                [$organizer],
                $this->broker->newConversationOptions->setPublic()
            );

            $event->addLocation($converstion->getAbsoluteUrl());
        }

        // ...
    }
}

```

### CODE REVIEW

File: lib/public/Talk/ITalkBackend.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+interface ITalkBackend {
+
+	/**
+	 * @param string $name
+	 * @param IConversationOptions|null $options optional configuration for the conversation
+	 *
+	 * @return IConversation
+	 * @since 24.0.0
+	 */
+	public function createPublicConversation(string $name, IConversationOptions $options = null): IConversation;
```
Comment (nickvergessen) [inline lib/public/Talk/ITalkBackend.php:45]: Public or not should be one of the conversation options

Comment (ChristophWurst) [reply lib/public/Talk/ITalkBackend.php:45]: Can you create a private conversation *without* an owner?

Comment (nickvergessen) [reply lib/public/Talk/ITalkBackend.php:45]: It should not be possible to have a private nor public conversation without a moderator who is able to delete it?

Comment (ChristophWurst) [reply lib/public/Talk/ITalkBackend.php:45]: Guess I just misunderstood. If there is always one mandatory owner (and moderator) then I can move that into a required parameter of `createRoom` and public/private can be a flag on the options. Makes sense.

Comment (ChristophWurst) [reply lib/public/Talk/ITalkBackend.php:45]: Adjusted.

**Comment by nickvergessen** [comment]
Looks good in general once the CS is fixed :P

**Review by miaulalala — APPROVED** [review]
LGTM!

File: lib/private/Talk/Broker.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+				$this->hasBackend = true;
+			} catch (Throwable $e) {
+				$this->logger->error("Talk backend {class} could not be loaded: " . $e->getMessage(), [
+					'class' => $backendRegistration->getService(),
+					'exception' => $e,
+				]);
+			}
+		}
+
+		return $this->hasBackend;
```
Comment (nickvergessen) [inline lib/private/Talk/Broker.php:90]: > lib/private/Talk/Broker.php:90:10: NullableReturnStatement: The declared return type 'bool' for OC\Talk\Broker::hasBackend is not nullable, but the function returns 'bool|null' (see https://psalm.dev/139)

I guess the try-catch block in the catch needs to set it to false?

Comment (ChristophWurst) [reply lib/private/Talk/Broker.php:90]: yes I have to restructure this

File: lib/public/Talk/ITalkBackend.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	 * @param string $name
+	 * @param IUser[] $moderators
+	 * @param IConversationOptions|null $options optional configuration for the conversation
+	 *
+	 * @return IConversation
+	 * @since 24.0.0
+	 */
+	public function createConversation(string $name,
+									   array $moderators = [],
+									   IConversationOptions $options = null): IConversation;
```
Comment (nickvergessen) [inline lib/public/Talk/ITalkBackend.php:51]: can we make this non option?
Or what is expected in that case, should we do "rand" for public or not?

File: lib/public/Talk/IConversationOptions.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+	/**
+	 * Make the new conversation public
+	 *
+	 * @param bool $isPublic
+	 *
+	 * @return $this
+	 * @since 24.0.0
+	 */
+	public function setPublic(bool $isPublic = true): self;
```
Comment (nickvergessen) [inline lib/public/Talk/IConversationOptions.php:41]: Needs a getter too, so Talk can get the details without assuming private api

**Comment by nickvergessen** [comment]
Talk draft implementation at https://github.com/nextcloud/spreed/pull/6858

File: tests/lib/Talk/BrokerTest.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			->method('getRegistrationContext')
+			->willReturn($this->createMock(RegistrationContext::class));
+
+		self::assertFalse(
+			$this->broker->hasBackend()
+		);
+	}
+
+	public function testHasFaultyBackend(): void {
+		$fakeTalkServiceClass = '\\OCA\\Spreed\\TalkBackend';
```
Comment (nickvergessen) [inline tests/lib/Talk/BrokerTest.php:80]: Use `Test` root namespace to make sure it never conflicts?

File: lib/private/Talk/ConversationOptions.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+namespace OC\Talk;
+
+use OCP\Talk\IConversationOptions;
+
+class ConversationOptions implements IConversationOptions {
+
+	/** @var bool */
+	private $isPublic;
```
Comment (nickvergessen) [inline lib/private/Talk/ConversationOptions.php:33]:
```suggestion
	private bool $isPublic;
```

File: lib/private/Talk/Broker.php [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	private $container;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var null|bool */
+	private $hasBackend = null;
+
+	/** @var null|ITalkBackend */
+	private $backend = null;
```
Comment (nickvergessen) [inline lib/private/Talk/Broker.php:54]:
```suggestion
	private Coordinator $coordinator;
	private IServerContainer $container;
	private LoggerInterface $logger;
	private ?bool $hasBackend;
	private ?ITalkBackend $backend;
```
?

File: tests/lib/Talk/BrokerTest.php (hunk 5 above)
Comment (ChristophWurst) [reply tests/lib/Talk/BrokerTest.php:80]: it's a unit test that doesn't perform any loading of sort. it should be fine

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**